File: color-palette-app/scripts/seed_database.py
from pymongo import MongoClient
import colorsys
import random
from datetime import datetime
import argparse

# Connexion à MongoDB
client = MongoClient('mongodb://localhost:27017/')
db = client['colorpalette']
collection = db['palettes']

# Liste des tags possibles
possible_tags = [
    'minimaliste', 'coloré', 'sombre', 'clair', 'pastel',
    'vintage', 'moderne', 'élégant', 'fun', 'professionnel',
    'artistique', 'chaleureux', 'froid', 'automne', 'printemps',
    'été', 'blanc'
]

def hsv_to_hex(h, s, v):
    """Convertit des valeurs HSV en code hexadécimal"""
    rgb = tuple(round(i * 255) for i in colorsys.hsv_to_rgb(h, s, v))
    return '#{:02x}{:02x}{:02x}'.format(*rgb)

def get_random_tags(count):
    """Retourne un nombre spécifique de tags aléatoires"""
    return random.sample(possible_tags, count)

def generate_monochromatic():
    """Génère une palette monochromatique"""
    hue = random.random()
    colors = []
    for i in range(5):
        value = 0.2 + (i * 0.15)  # Variation de la luminosité
        colors.append(hsv_to_hex(hue, 0.6, value))
    return {
        'colors': colors,
        'tags': ['monochrome'] + get_random_tags(1),
        'createdAt': datetime.utcnow()
    }

def generate_analogous():
    """Génère une palette de couleurs analogues"""
    base_hue = random.random()
    colors = []
    for i in range(-2, 3):
        hue = (base_hue + (i * 0.083)) % 1  # 30 degrés en proportion (30/360)
        colors.append(hsv_to_hex(hue, 0.6, 0.7))
    return {
        'colors': colors,
        'tags': ['analogue'] + get_random_tags(1),
        'createdAt': datetime.utcnow()
    }

def generate_complementary():
    """Génère une palette de couleurs complémentaires"""
    base_hue = random.random()
    complement_hue = (base_hue + 0.5) % 1
    colors = []
    
    # Couleur de base et variations
    for i in range(3):
        value = 0.3 + (i * 0.2)
        colors.append(hsv_to_hex(base_hue, 0.6, value))
    
    # Couleur complémentaire et variation
    for i in range(2):
        value = 0.4 + (i * 0.2)
        colors.append(hsv_to_hex(complement_hue, 0.6, value))
    
    return {
        'colors': colors,
        'tags': ['complémentaire'] + get_random_tags(1),
        'createdAt': datetime.utcnow()
    }

def generate_triadic():
    """Génère une palette de type triade"""
    base_hue = random.random()
    hue2 = (base_hue + 0.333) % 1
    hue3 = (base_hue + 0.666) % 1
    colors = []
    
    # Une couleur de chaque triade avec variations
    for hue in [base_hue, hue2, hue3]:
        for i in range(2):
            value = 0.4 + (i * 0.2)
            colors.append(hsv_to_hex(hue, 0.6, value))
    
    return {
        'colors': colors[:5],  # On garde seulement 5 couleurs maximum
        'tags': ['triade'] + get_random_tags(1),
        'createdAt': datetime.utcnow()
    }

def generate_random_palette():
    """Génère une palette aléatoire selon une des règles"""
    generators = [
        generate_monochromatic,
        generate_analogous,
        generate_complementary,
        generate_triadic
    ]
    return random.choice(generators)()

def seed_database(nb_palettes):
    """Ajoute le nombre spécifié de palettes à la base de données"""
    try:
        # Génère les nouvelles palettes
        palettes = [generate_random_palette() for _ in range(nb_palettes)]
        
        # Ajoute les nouvelles palettes
        collection.insert_many(palettes)
        print(f'{nb_palettes} nouvelles palettes ont été ajoutées à la base de données')

    except Exception as e:
        print('Erreur lors du seed de la base de données:', str(e))
    finally:
        client.close()

if __name__ == '__main__':
    # Configuration du parser d'arguments
    parser = argparse.ArgumentParser(description='Génère des palettes de couleurs harmonieuses')
    parser.add_argument('--nombre', '-n', type=int, default=20,
                      help='Nombre de palettes à générer (défaut: 20)')
    
    # Récupération des arguments
    args = parser.parse_args()
    
    # Lancement du seed avec le nombre spécifié
    seed_database(args.nombre)

File: color-palette-app/scripts/seed_database.js
const { MongoClient } = require('mongodb');
const yargs = require('yargs');

// Liste des tags possibles
const possibleTags = [
    'minimaliste', 'coloré', 'sombre', 'clair', 'pastel',
    'vintage', 'moderne', 'élégant', 'fun', 'professionnel',
    'artistique', 'chaleureux', 'froid', 'automne', 'printemps',
    'été', 'blanc'
];

// Fonction pour convertir HSV en RGB puis en Hex
function hsvToHex(h, s, v) {
    let r, g, b;
    
    const i = Math.floor(h * 6);
    const f = h * 6 - i;
    const p = v * (1 - s);
    const q = v * (1 - f * s);
    const t = v * (1 - (1 - f) * s);

    switch (i % 6) {
        case 0: r = v; g = t; b = p; break;
        case 1: r = q; g = v; b = p; break;
        case 2: r = p; g = v; b = t; break;
        case 3: r = p; g = q; b = v; break;
        case 4: r = t; g = p; b = v; break;
        case 5: r = v; g = p; b = q; break;
    }

    const toHex = x => {
        const hex = Math.round(x * 255).toString(16);
        return hex.length === 1 ? '0' + hex : hex;
    };

    return `#${toHex(r)}${toHex(g)}${toHex(b)}`;
}

// Fonction pour obtenir des tags aléatoires
function getRandomTags(count) {
    const shuffled = [...possibleTags].sort(() => 0.5 - Math.random());
    return shuffled.slice(0, count);
}

// Générateurs de palettes
function generateMonochromatic() {
    const hue = Math.random();
    const colors = Array.from({ length: 5 }, (_, i) => {
        const value = 0.2 + (i * 0.15);
        return hsvToHex(hue, 0.6, value);
    });

    return {
        colors,
        tags: ['monochrome', ...getRandomTags(1)],
        createdAt: new Date()
    };
}

function generateAnalogous() {
    const baseHue = Math.random();
    const colors = Array.from({ length: 5 }, (_, i) => {
        const hue = (baseHue + ((i - 2) * 0.083)) % 1;
        return hsvToHex(hue, 0.6, 0.7);
    });

    return {
        colors,
        tags: ['analogue', ...getRandomTags(1)],
        createdAt: new Date()
    };
}

function generateComplementary() {
    const baseHue = Math.random();
    const complementHue = (baseHue + 0.5) % 1;
    const colors = [
        ...Array.from({ length: 3 }, (_, i) => {
            const value = 0.3 + (i * 0.2);
            return hsvToHex(baseHue, 0.6, value);
        }),
        ...Array.from({ length: 2 }, (_, i) => {
            const value = 0.4 + (i * 0.2);
            return hsvToHex(complementHue, 0.6, value);
        })
    ];

    return {
        colors,
        tags: ['complémentaire', ...getRandomTags(1)],
        createdAt: new Date()
    };
}

function generateTriadic() {
    const baseHue = Math.random();
    const hue2 = (baseHue + 0.333) % 1;
    const hue3 = (baseHue + 0.666) % 1;
    
    const colors = [baseHue, hue2, hue3].flatMap(hue => 
        Array.from({ length: 2 }, (_, i) => {
            const value = 0.4 + (i * 0.2);
            return hsvToHex(hue, 0.6, value);
        })
    ).slice(0, 5);

    return {
        colors,
        tags: ['triade', ...getRandomTags(1)],
        createdAt: new Date()
    };
}

function generateRandomPalette() {
    const generators = [
        generateMonochromatic,
        generateAnalogous,
        generateComplementary,
        generateTriadic
    ];
    const randomGenerator = generators[Math.floor(Math.random() * generators.length)];
    return randomGenerator();
}

async function seedDatabase(nbPalettes) {
    const uri = 'mongodb://localhost:27017';
    const client = new MongoClient(uri);

    try {
        await client.connect();
        console.log('Connecté à MongoDB');

        const db = client.db('colorpalette');
        const collection = db.collection('palettes');

        // Génère les nouvelles palettes
        const palettes = Array.from({ length: nbPalettes }, () => generateRandomPalette());
        
        // Ajoute les nouvelles palettes
        await collection.insertMany(palettes);
        console.log(`${nbPalettes} nouvelles palettes ont été ajoutées à la base de données`);

    } catch (error) {
        console.error('Erreur lors du seed de la base de données:', error);
    } finally {
        await client.close();
        console.log('Déconnecté de MongoDB');
    }
}

// Configuration des arguments en ligne de commande
const argv = yargs
    .option('nombre', {
        alias: 'n',
        description: 'Nombre de palettes à générer',
        type: 'number',
        default: 20
    })
    .help()
    .argv;

// Lancement du script
seedDatabase(argv.nombre);
